generator/internal/discovery_type_vertex.cc

```diff
@@ -172,87 +183,113 @@ auto constexpr kMaxRecursionDepth = 32;
 // NOLINTNEXTLINE(misc-no-recursion)
 StatusOr<std::string> DiscoveryTypeVertex::FormatMessage(
     std::map<std::string, DiscoveryTypeVertex> const& types,
-    std::string const& name, std::string const& file_package_name,
-    nlohmann::json const& json, int indent_level) const {
+    std::string const& name, std::string const& qualified_name,
+    std::string const& file_package_name, nlohmann::json const& json,
+    int indent_level) const {
   if (indent_level > kMaxRecursionDepth) {
     GCP_LOG(FATAL) << __func__ << " exceeded kMaxRecursionDepth";
   }
   std::string indent(indent_level * 2, ' ');
-  auto lines =
-      FormatProperties(types, name, file_package_name, json, indent_level + 1);
-  if (!lines) return std::move(lines).status();
-  return absl::StrCat(indent, absl::StrFormat("message %s {\n", name),
-                      absl::StrJoin(*lines, "\n\n"), "\n", indent, "}");
+  auto properties = FormatProperties(types, name, qualified_name,
+                                     file_package_name, json, indent_level + 1);
+  if (!properties) return std::move(properties).status();
+  std::vector<std::string> message_name_parts =
+      absl::StrSplit(name, absl::ByChar('.'));
+  std::string reserved_numbers;
+  if (!properties->reserved_numbers.empty()) {
+    reserved_numbers =
+        absl::StrCat(indent, "  reserved ",
+                     absl::StrJoin(properties->reserved_numbers, ", "), ";\n");
+  }
+  return absl::StrCat(
+      indent, absl::StrFormat("message %s {\n", message_name_parts.back()),
+      reserved_numbers, absl::StrJoin(properties->lines, "\n\n"), "\n", indent,
+      "}");
 }
 
 DiscoveryTypeVertex::MessageProperties
 DiscoveryTypeVertex::DetermineReservedAndMaxFieldNumbers(
     google::protobuf::Descriptor const& message_descriptor) {
-  MessageProperties message_properties{{}, 0};
+  MessageProperties message_properties{{}, {}, 0};
   for (auto r = 0; r != message_descriptor.reserved_range_count(); ++r) {
     auto const* reserved_range = message_descriptor.reserved_range(r);
     for (int j = reserved_range->start; j < reserved_range->end; ++j) {
       message_properties.reserved_numbers.insert(j);
     }
-    // google::protobuf::ReservedRange.end returns the next available value, not
-    // the actual end reserved value.
     message_properties.next_available_field_number =
         std::max(message_properties.next_available_field_number,
                  message_descriptor.reserved_range(r)->end);
   }
-
   for (auto i = 0; i != message_descriptor.field_count(); ++i) {
     message_properties.next_available_field_number =
         std::max(message_properties.next_available_field_number,
                  message_descriptor.field(i)->number() + 1);
   }
-
   return message_properties;
 }
 
-// NOLINTNEXTLINE(misc-no-recursion)
-StatusOr<std::vector<std::string>> DiscoveryTypeVertex::FormatProperties(
+// TODO(#12234): Refactor this function into multiple smaller functions to
```

Comment:
The sooner the better, this is already too big.  Maybe split the loop to return `current_field_names`?